### PR TITLE
threads: add remaining instructions

### PR DESCRIPTION
**Pull request by abrown**
With #1649 merged, we can now implement the remaining shared-everything-threads [instructions]. Many of the tests are generated, so there may be tweaks coming in the future; this change just gets the initial encoding, decoding, and validation pieces in place.

[instructions]: https://github.com/WebAssembly/shared-everything-threads/blob/main/proposals/shared-everything-threads/Overview.md#instructions

### CODE REVIEW

File: crates/wasmparser/src/validator/operators.rs [hunk 1]
```diff
@@ -1155,6 +1144,40 @@ where
...
+    /// Common helper for checking the types of arrays accessed with atomic RMW
+    /// instructions, which only allow `i32` and `i64`.
+    fn check_atomic_array_rmw_ty(&self, type_index: u32) -> Result<ValType> {
+        let elem_ty = self.array_type_at(type_index)?.0.element_type;
+        match elem_ty {
+            StorageType::Val(ValType::I32) => Ok(ValType::I32),
+            StorageType::Val(ValType::I64) => Ok(ValType::I64),
+            _ => bail!(
+                self.offset,
+                "invalid type: `struct.atomic.rmw.*` only allows `i32` and `i64`"
```
Comment (alexcrichton) [inline crates/wasmparser/src/validator/operators.rs:1176]: copy/paste with `struct.atomic.*` here, but perhaps this and the above method could share more implementation details?

File: crates/wasmparser/src/validator/operators.rs [hunk 2]
```diff
@@ -3757,6 +3902,149 @@ where
...
+        _ordering: Ordering,
+        struct_type_index: u32,
+        field_index: u32,
+    ) -> Self::Output {
+        let field_ty = self.check_atomic_struct_rmw_ty(struct_type_index, field_index)?;
+        self.pop_operand(Some(field_ty))?;
+        self.pop_concrete_ref(true, struct_type_index)?;
+        self.push_operand(field_ty)?;
+        Ok(())
+    }
```
Comment (alexcrichton) [inline crates/wasmparser/src/validator/operators.rs:3990]: Could these method bodies be refactored into a shared function?

File: crates/wasmparser/src/validator/operators.rs [hunk 3]
```diff
@@ -3989,6 +4333,99 @@ where
...
+    fn visit_array_atomic_rmw_xor(&mut self, _ordering: Ordering, type_index: u32) -> Self::Output {
+        let elem_ty = self.check_atomic_array_rmw_ty(type_index)?;
+        self.pop_operand(Some(elem_ty))?;
+        self.pop_operand(Some(ValType::I32))?;
+        self.pop_concrete_ref(true, type_index)?;
+        self.push_operand(elem_ty)?;
+        Ok(())
+    }
```
Comment (alexcrichton) [inline crates/wasmparser/src/validator/operators.rs:4375]: Like structs could these be moved into a shared function?

File: crates/wasmparser/src/validator/operators.rs (hunk 1 above)
Comment (abrown) [reply crates/wasmparser/src/validator/operators.rs]: Fixed the typo and looked at merging them: the only part that is the same is the i32/i64 check. The other parts--type access and push/pop logic--ends up being different.

**Pull request merged**